src/Microsoft.FeatureManagement/FeatureManagerSnapshot.cs

```diff
@@ -43,34 +45,26 @@ await foreach (string featureName in _featureManager.GetFeatureNamesAsync().Conf
 
         public async Task<bool> IsEnabledAsync(string feature)
         {
-            //
-            // First, check local cache
-            if (_flagCache.ContainsKey(feature))
-            {
-                return _flagCache[feature];
-            }
-
-            bool enabled = await _featureManager.IsEnabledAsync(feature).ConfigureAwait(false);
-
-            _flagCache[feature] = enabled;
+            Lazy<Task<bool>> evaluator = _flagCache.GetOrAdd(
+                feature,
+                (key) => 
+                    new Lazy<Task<bool>>(
+                        () => _featureManager.IsEnabledAsync(key),
+                        LazyThreadSafetyMode.ExecutionAndPublication));
 
-            return enabled;
+            return await evaluator.Value.ConfigureAwait(false);
```

Comment:
You don't need to await the task here, just return it thus saving the async state machinery (allocations).

I'm not sure if the `Lazy` is needed. It's safer for race conditions in the sense of flag evaluations, but caching is a best effort solution which should be fast, so I would trade perf over possible "double evaluations".

Reply:
@gfoidl thanks that's a good improvement with the await removal.

I'm going to remove the double evaluation protection in this PR and think about that problem a bit more.